File: scheduler_py/soft_constraints.py
from typing import List, Dict, Any
from .models import ActivityArea, Cabin, Period, Assignment
from .utils import (
    get_last_cabin_area,
    calculate_travel_time,
    get_area_utilization,
)

def rank_candidate_areas(
    candidate_areas: List[ActivityArea],
    cabin: Cabin,
    period: Period,
    assignments: List[Assignment],
    config: Dict[str, Any],
    all_periods: List[Period]
) -> List[ActivityArea]:
    """Rank candidate areas based on soft constraints."""
    scored_areas = [
        (area, calculate_area_score(area, cabin, period, assignments, config, all_periods))
        for area in candidate_areas
    ]
    scored_areas.sort(key=lambda x: x[1], reverse=True)
    return [area for area, score in scored_areas]

def calculate_area_score(
    area: ActivityArea,
    cabin: Cabin,
    period: Period,
    assignments: List[Assignment],
    config: Dict[str, Any],
    all_periods: List[Period]
) -> float:
    """Calculate a score for an area based on soft constraints."""
    score = 100.0  # Base score

    score += calculate_age_group_priority_score(area, cabin, config)
    score += calculate_area_variety_score(area, cabin, period, assignments, config)
    score += calculate_social_grouping_score(area, cabin, period, assignments)
    score += calculate_preference_score(area, cabin)
    score += calculate_travel_time_score(area, cabin, period, assignments, config, all_periods)
    score += calculate_utilization_goal_score(area, period, assignments, config)
    score += calculate_weather_score(area, config)

    return max(0, score)

def calculate_age_group_priority_score(area: ActivityArea, cabin: Cabin, config: Dict[str, Any]) -> float:
    """Calculate score based on age group priority."""
    age_group_priorities = config.get("ageGroupPriorities", [])
    for priority in age_group_priorities:
        if priority.get("ageGroup") == cabin.age_group and priority.get("areaId") == area.id:
            return priority.get("priority", 0) * 10
    return 0

def calculate_area_variety_score(area: ActivityArea, cabin: Cabin, period: Period, assignments: List[Assignment], config: Dict[str, Any]) -> float:
    """Calculate score based on area variety."""
    score = 0.0
    cabin_assignments = [a for a in assignments if a.cabin_id == cabin.id]
    recent_assignments = [a for a in cabin_assignments if period.day - 2 <= a.day <= period.day]

    if recent_assignments:
        last_assignment = recent_assignments[-1]
        last_area = next((a for a in config.get("areas", []) if a.id == last_assignment.area_id), None)
        if last_area and last_area.category == area.category:
            score -= 20

    used_categories = {
        next((a.category for a in config.get("areas", []) if a.id == asgn.area_id), None)
        for asgn in recent_assignments
    }
    if area.category not in used_categories:
        score += 15

    return score

def calculate_social_grouping_score(area: ActivityArea, cabin: Cabin, period: Period, assignments: List[Assignment]) -> float:
    """Calculate score based on social grouping preferences."""
    if not cabin.social_groups:
        return 0.0

    score = 0.0
    current_period_assignments = [a for a in assignments if a.day == period.day and a.period_id == period.id]

    for social_group_id in cabin.social_groups:
        if any(a.cabin_id == social_group_id and a.area_id == area.id for a in current_period_assignments):
            score += 25

    return score

def calculate_preference_score(area: ActivityArea, cabin: Cabin) -> float:
    """Calculate score based on cabin preferences."""
    score = 0.0
    if area.id in cabin.preferences.favorite_areas:
        score += 30
    if area.id in cabin.preferences.avoid_areas:
        score -= 50
    return score

def calculate_travel_time_score(area: ActivityArea, cabin: Cabin, period: Period, assignments: List[Assignment], config: Dict[str, Any], all_periods: List[Period]) -> float:
    """Calculate score based on travel time optimization."""
    last_area_id = get_last_cabin_area(cabin.id, assignments, period.day, period.id, all_periods)
    if not last_area_id:
        return 0.0

    last_area = next((a for a in config.get("areas", []) if a.id == last_area_id), None)
    if not last_area:
        return 0.0

    travel_time = calculate_travel_time(last_area, area)
    max_allowed_time = config.get("allowedTransitionTime", 30)

    return 10 if travel_time <= max_allowed_time else -15

def calculate_utilization_goal_score(area: ActivityArea, period: Period, assignments: List[Assignment], config: Dict[str, Any]) -> float:
    """Calculate score based on area utilization goals."""
    utilization_goals = config.get("areaUtilizationGoals", [])
    goal = next((g for g in utilization_goals if g.get("areaId") == area.id), None)
    if not goal:
        return 0.0

    current_utilization = get_area_utilization(area.id, assignments, period.day, period.id)
    target_utilization = goal.get("targetUtilization", area.max_capacity * 0.8)

    if current_utilization < target_utilization:
        return 20
    elif current_utilization >= area.max_capacity:
        return -30
    return 0

def calculate_weather_score(area: ActivityArea, config: Dict[str, Any]) -> float:
    """Calculate score based on weather considerations."""
    if not area.weather_sensitive:
        return 0.0
    # Placeholder for weather integration
    return 0.0

# Cabin merging logic is not included in the soft constraints for ranking,
# but would be applied before the scheduling loop begins.
# These functions are placeholders for that logic.

def get_cabin_merge_instructions(cabin: Cabin, config: Dict[str, Any]) -> Any:
    """Check if cabin merging should be applied."""
    if not config.get("cabinMergingModel") or not config.get("mergeInstructions"):
        return None
    return next((inst for inst in config["mergeInstructions"] if inst.get("cabinId") == cabin.id), None)

def apply_cabin_merging(cabins: List[Cabin], config: Dict[str, Any]) -> List[Cabin]:
    """Apply cabin merging logic."""
    if config.get("cabinMergingModel") == "none":
        return cabins
    # Placeholder for more complex merging logic
    return cabins

File: src/algorithm/constraints/softConstraints.js
/**
 * Soft constraint functions for camp scheduling
 * These functions implement the "Conditionals" (soft rules) for ranking candidate areas
 */

import {
  getLastCabinArea,
  getAreaUtilization,
  hasCabinUsedAreaRecently
} from './utilities.js';

/**
 * Rank candidate areas based on soft constraints
 * @param {Array} candidateAreas - Array of available areas
 * @param {Object} cabin - Cabin object
 * @param {Object} period - Period object
 * @param {Array} assignments - Current assignments
 * @param {Object} config - Configuration object
 * @param {Array} flatAreas - Flattened array of all areas
 * @returns {Array} Sorted array of areas with scores
 */
export function rankCandidateAreas(candidateAreas, cabin, period, assignments, config, flatAreas) {
  const scoredAreas = candidateAreas.map(area => {
    const score = calculateAreaScore(area, cabin, period, assignments, config, flatAreas);
    return { area, score };
  });

  // Sort by score (higher is better) and return areas
  return scoredAreas
    .sort((a, b) => b.score - a.score)
    .map(item => item.area);
}

/**
 * Calculate a score for an area based on soft constraints
 * @param {Object} area - Activity area
 * @param {Object} cabin - Cabin object
 * @param {Object} period - Period object
 * @param {Array} assignments - Current assignments
 * @param {Object} config - Configuration object
 * @param {Array} flatAreas - Flattened array of all areas
 * @returns {number} Score (higher is better)
 */
function calculateAreaScore(area, cabin, period, assignments, config, flatAreas) {
  let score = 100; // Base score

  // Age group priority scoring
  score += calculateAgeGroupPriorityScore(area, cabin, config);

  // Area variety scoring
  score += calculateAreaVarietyScore(area, cabin, period, assignments, config, flatAreas);

  // Social grouping scoring
  score += calculateSocialGroupingScore(area, cabin, period, assignments, config);

  // Preference scoring
  score += calculatePreferenceScore(area, cabin);

  // Area utilization goals
  score += calculateUtilizationGoalScore(area, period, assignments, config);

  // Weather considerations
  score += calculateWeatherScore(area, config);

  return Math.max(0, score); // Ensure non-negative score
}

/**
 * Calculate score based on age group priority
 * @param {Object} area - Activity area
 * @param {Object} cabin - Cabin object
 * @param {Object} config - Configuration object
 * @returns {number} Score adjustment
 */
function calculateAgeGroupPriorityScore(area, cabin, config) {
  if (!config.ageGroupPriorities) return 0;

  const priority = config.ageGroupPriorities.find(p =>
    p.ageGroup === cabin.ageGroup && p.areaId === area.id
  );

  if (priority) {
    return priority.priority * 10; // Multiply by 10 for significant impact
  }

  return 0;
}

/**
 * Calculate score based on area variety
 * @param {Object} area - Activity area
 * @param {Object} cabin - Cabin object
 * @param {Object} period - Period object
 * @param {Array} assignments - Current assignments
 * @param {Object} config - Configuration object
 * @param {Array} flatAreas - Flattened array of all areas
 * @returns {number} Score adjustment
 */
function calculateAreaVarietyScore(area, cabin, period, assignments, config, flatAreas) {
  let score = 0;

  // Get cabin's recent area history
  const cabinAssignments = assignments.filter(a => a.cabinId === cabin.id);
  const recentAssignments = cabinAssignments.filter(a =>
    a.day >= period.day - 2 && a.day <= period.day
  );

  // Check for back-to-back similar categories
  const lastAssignment = recentAssignments[recentAssignments.length - 1];
  if (lastAssignment) {
    const lastArea = flatAreas.find(a => a.id === lastAssignment.areaId);
    if (lastArea && lastArea.category === area.category) {
      score -= 20; // Penalty for consecutive similar activities
    }
  }

  // Bonus for variety
  const usedCategories = new Set(recentAssignments.map(assignment => {
    const assignedArea = flatAreas.find(ar => ar.id === assignment.areaId);
    return assignedArea ? assignedArea.category : null;
  }).filter(Boolean));

  if (!usedCategories.has(area.category)) {
    score += 15; // Bonus for new category
  }

  return score;
}

/**
 * Calculate score based on social grouping preferences
 * @param {Object} area - Activity area
 * @param {Object} cabin - Cabin object
 * @param {Object} period - Period object
 * @param {Array} assignments - Current assignments
 * @param {Object} config - Configuration object
 * @returns {number} Score adjustment
 */
function calculateSocialGroupingScore(area, cabin, period, assignments, config) {
  if (!cabin.socialGroups || cabin.socialGroups.length === 0) {
    return 0;
  }

  let score = 0;
  const currentPeriodAssignments = assignments.filter(a =>
    a.day === period.day && a.periodId === period.id
  );

  // Check if social group cabins are assigned to this area
  for (const socialGroupId of cabin.socialGroups) {
    const socialCabinAssignment = currentPeriodAssignments.find(a =>
      a.cabinId === socialGroupId && a.areaId === area.id
    );

    if (socialCabinAssignment) {
      score += 25; // Bonus for being with social group
    }
  }

  return score;
}

/**
 * Calculate score based on cabin preferences
 * @param {Object} area - Activity area
 * @param {Object} cabin - Cabin object
 * @returns {number} Score adjustment
 */
function calculatePreferenceScore(area, cabin) {
  let score = 0;

  if (cabin.preferences) {
    // Favorite areas bonus
    if (cabin.preferences.favoriteAreas &&
        cabin.preferences.favoriteAreas.includes(area.id)) {
      score += 30;
    }

    // Avoid areas penalty
    if (cabin.preferences.avoidAreas &&
        cabin.preferences.avoidAreas.includes(area.id)) {
      score -= 50;
    }
  }

  return score;
}

/**
 * Calculate score based on area utilization goals
 * @param {Object} area - Activity area
 * @param {Object} period - Period object
 * @param {Array} assignments - Current assignments
 * @param {Object} config - Configuration object
 * @returns {number} Score adjustment
 */
function calculateUtilizationGoalScore(area, period, assignments, config) {
  if (!config.areaUtilizationGoals) return 0;

  const goal = config.areaUtilizationGoals.find(g => g.areaId === area.id);
  if (!goal) return 0;

  const currentUtilization = getAreaUtilization(area.id, assignments, period.day, period.id);
  const targetUtilization = goal.targetUtilization || area.maxCapacity * 0.8;

  if (currentUtilization < targetUtilization) {
    return 20; // Bonus for helping meet utilization goals
  } else if (currentUtilization >= area.maxCapacity) {
    return -30; // Penalty for over-utilization
  }

  return 0;
}

/**
 * Calculate score based on weather considerations
 * @param {Object} area - Activity area
 * @param {Object} config - Configuration object
 * @returns {number} Score adjustment
 */
function calculateWeatherScore(area, config) {
  if (!area.weatherSensitive) return 0;

  // This would typically integrate with actual weather data
  // For now, return a neutral score
  return 0;
}

/**
 * Check if cabin merging should be applied
 * @param {Object} cabin - Cabin object
 * @param {Object} config - Configuration object
 * @returns {Object|null} Merge instructions or null
 */
export function getCabinMergeInstructions(cabin, config) {
  if (!config.cabinMergingModel || !config.mergeInstructions) {
    return null;
  }

  return config.mergeInstructions.find(instruction =>
    instruction.cabinId === cabin.id
  );
}

/**
 * Apply cabin merging logic
 * @param {Array} cabins - Array of cabins
 * @param {Object} config - Configuration object
 * @returns {Array} Processed cabins with merging applied
 */
export function applyCabinMerging(cabins, config) {
  if (!config.cabinMergingModel || config.cabinMergingModel === 'none') {
    return cabins;
  }

  // This is a placeholder for cabin merging logic
  // Implementation would depend on the specific merging model
  return cabins;
}
